### ISSUE TITLE
RAM not freed after ruperemove is killed/cancelled

### ISSUE THREAD

**Issue by Enverex** [markfasheh/duperemove#42]
If duperemove is killed (due to OOM) or cancelled (via Ctrl+C) it doesn't appear to free the majority of the RAM it was using.

Example:
System had ~7600MB available (of 7950+MB) after boot.
Ran "duperemove --read-hashes=dedupe.hashes -dh" and hit Ctrl+C to cancel when system reached 650MB available RAM to avoid an OOM situation due to too many files/hashes.
System is now down to 2200MB RAM free after duperemove has closed.
"top" only shows processes using very little RAM which were there before duperemove was started and only account for the 300MB or so that was originally in use (SlabTop shows the kernel is only using around 230MB).

Somehow nearly 6GB of RAM is now unaccounted for.

Is this potentially a memory leak in duperemove, btrfs or the kernel?

**Comment by clobrother** [comment]
Are not you mistaken about the memory cache ?
Can you post a "free -m" output ?

**Comment by Enverex** [comment]
I'm definitely not mistaken, I'm not a Linux newbie ;)

I was comparing the system's available RAM (excluding buffers/cache) before and after.

I've just tried to reproduce this so I can show you the exact figures but I didn't cancel it soon enough and it's run out of RAM entirely and I think the machine has kernel panic'd as I can't reach it any more so I'm going to have to go and hard reboot it (I'm not in the same location).

**Comment by Enverex** [comment]
Here are the specifics:

Before running "duperemove":

              total        used        free      shared  buff/cache   available
Mem:           7909         362        7252           7         294        7318
Swap:             0           0           0

Cancelled after 5 minutes of running:

              total        used        free      shared  buff/cache   available
Mem:           7909        2187        2509           7        3212        5390
Swap:             0           0           0


As you can see the system now only has 5.4GB usable, rather than the ~7.3GB I'd expect. "px aux" and "top" don't show any additional processes using this extra RAM, the process list looks the same as it did before duperemove was ran yet a large amount of RAM is now missing.

**Comment by kakra** [comment]
Does your kernel use HugeTLB or transparent huge pages? It may have an impact... I'd turn that off for normal server and desktop usage. It is only interesting if you handle huge amounts of data, e.g. for calculations.

**Comment by dioni21** [comment]
Run top, order by RSS, and observe variation while running and after canceling dupremove.

**Comment by Enverex** [comment]
@kakra Not unless it's a normal feature, I'm using the stock Arch Linux kernel on that box.

@dioni21 Nothing other than duperemove shows any real movement there. duperemove is currently showing 811MB VIRT and 687MB RES/RSS but in reality the system is currently at 3.1GB available (which means duperemove or BTRFS is using 4.2GB).

**Comment by markfasheh** [comment]
Hi, thanks for reporting this. A couple questions:

- What kernel version are you running?

- Also, what version of duperemove are you running? Master or v0.09-branch (or something else)?

Btw, if you're running master and don't mind checking whether v0.09-branch exhibits the same behavior, I'd appreciate it (though it wouldn't be strictly necessary to figuring this out).

So if you're seeing memory around after the process is killed than that indicates a kernel leak. At what point in the process do you kill duperemove to get this result?

Basically if you're killing it in the csum or extent search phase, and can observe the leak we'd want to look at the btrfs calls we make there. Otherwise, it's probably in the dedupe stage which is mostly fiemap and calls to the btrfs-extent-same ioctl.

On my end, I actually saw some issues running many threads of dedupe on master branch shortly after committing my multi-threaded dedupe work, but didn't have time to figure out exactly what was happening there.

**Comment by Enverex** [comment]
Kernel is 3.17.6

I'm running duperemove from GIT at commit number 238 (2435a6e - which is yesterday I believe).

It happens during the scanning (csum) stage from what I can tell. If I start the scan and cancel it (Ctrl+C) then I don't seem to get my RAM back. I have to reboot to get back to normal.

Let me know if you want me to try anything specific.

**Comment by markfasheh** [comment]
Ok, so if you're experiencing this in the scan phase and you've verified that all duperemove processes are killed then we're be looking at the btrfs specific calls made during that phase (which isn't a whole lot, really).

- try with v0.09-branch so we know whether this was a recent change in the duperemove code (pls attach maybe the first 10 lines of output so we can verify that it's running the right version, etc)

- Try against a different file system so we can know whether this is related to btrfs or not. If this isn't possible try with this branch (though really, if you can try a different fs altogether that's best):
    https://github.com/markfasheh/duperemove/tree/no-btrfs

**Comment by dioni21** [comment]
Quick and dumb questions:
@markfasheh - Do you use any kind of shared memory in dupremove? (could be useful in parallel scans)
@Enverex - Compare the output of the command "ipcs", before and after "losing" memory

**Comment by Enverex** [comment]
@markfasheh Duperemove reports "duperemove v0.10-dev". I lose the RAM regardless which stage it fails at it seems. I just tried to do another (too large) folder and it OOMed on the dedupe stage and I didn't get the RAM back.

@dioni21 Before and after:

	------ Message Queues --------
	key        msqid      owner      perms      used-bytes   messages

	------ Shared Memory Segments --------
	key        shmid      owner      perms      bytes      nattch     status

	------ Semaphore Arrays --------
	key        semid      owner      perms      nsems
	0x00000000 98304      http       600        1
	0x00000000 131073     http       600        1
	0x00000000 65538      http       600        1
	0x00000000 163843     http       600        1

After:

	------ Message Queues --------
	key        msqid      owner      perms      used-bytes   messages

	------ Shared Memory Segments --------
	key        shmid      owner      perms      bytes      nattch     status
	0x00000000 0          root       600        393216     2          dest
	0x00000000 32769      root       600        196608     2          dest
	0x00000000 65538      root       600        393216     2          dest
	0x00000000 98307      root       600        393216     2          dest
	0x00000000 131076     root       600        393216     2          dest
	0x00000000 229381     root       600        196608     2          dest
	0x00000000 327686     root       600        393216     2          dest
	0x00000000 557063     root       600        393216     2          dest
	0x00000000 655368     root       600        524288     2          dest

	------ Semaphore Arrays --------
	key        semid      owner      perms      nsems
	0x00000000 98304      http       600        1
	0x00000000 131073     http       600        1
	0x00000000 65538      http       600        1
	0x00000000 163843     http       600        1

**Comment by dioni21** [comment]
@Enverex You can probably get back your memory with the **`ipcrm`** command:

**`ipcrm --all=shm`**

**Comment by dioni21** [comment]
BTW: Your shared memory apparently still has some process attached.  To determine which, read this: http://stackoverflow.com/questions/5658568/how-to-list-processes-attached-to-a-shared-memory-segment-in-linux

**Comment by ariccio** [comment]
I'm sorry, but I can't help it - "ruperemove"
![image](https://cloud.githubusercontent.com/assets/2142308/5865107/1b3c1420-a255-11e4-80ac-472f40509198.png)

**Comment by clobrother** [comment]
I reached this issue too
I am using the kernel 3.16, last duperemove code from master
There is no shared memory in use (or few)

The issue seems to be reproductible
On the dedup phase, memory grows, duperemove is killed by oomk, memory is not freed
Note that my swap is not used (I guess kernel space do not use swap, indicating the issue is in kernel space)
Note also that the issue is not reached when I am deduping with a single thread : maybe the threading cause the leak, or maybe, with a single thread, oomk does not kill me (showing an issue in the error code path, maybe)

**Comment by markfasheh** [comment]
Ok this is exactly what I have seen with master. Basically after I threaded the dedupe stage I could get memory would grow until it oom killer kicks in. This doesn't happen in v0.09.  Running dedupe with one thread 'fixes' it back too.

From what I can tell, either we're leaking in duperemove or the kernel is keeping some data pinned longer than it needs to. I'm gonna mark this as a bug that should be fixed before v0.10 release (well either that or we roll back the threaded dedupe code).

**Comment by dioni21** [comment]
Did you try the ipcrm command I suggested above?

SYSV IPC(2) data is kept if the program crashes.  @markfasheh , If you use shmget(2), try using BSD versions of IPCs, like mmap(2).  These are freed if the program crashes.

**Comment by markfasheh** [comment]
I'll give it a shot again in a bit. We don't use shared memory however I wonder if glib2 is using it underneath us.

**Comment by dioni21** [comment]
Anyway, "ipcs -p" or even "strace -f" would help debug.   ;-)

**Comment by clobrother** [comment]
There is no shared memory lurking after the kill (only a few kb, used by my browser or something)

**Comment by dioni21** [comment]
@clobrother how do you know that memory has not been freed?  If you can reproduce the error, could you please show us the output for "top -Sbn1 -o RES | head -30" before and after the memory leak?

**Comment by clobrother** [comment]
@dioni21 I've almost no memory used in normal usage (somethink like 300Mb/4GB)
After dedup, 3800MB was used (no cache, shown as "green" in htop

Killing all usedspace program didn't change that usage

**Comment by dioni21** [comment]
@clobrother In top, "used" memory means not much.  It could be used by buffers AND/OR cached.

Some refs:
http://www.linuxatemyram.com/
http://unix.stackexchange.com/questions/87908/how-do-you-empty-the-buffers-and-cache-on-a-linux-system

Try cleaning the cache, like said on the second ref: sudo sysctl vm.drop_caches=3

If this helps, this is no bug!  Just Linux working...

**Comment by clobrother** [comment]
@dioni21 there are some data:
Before:
```
13% [jack:~]top -Sbn1 -o RES | head -30
top - 23:25:34 up  5:15,  6 users,  load average: 0,91, 1,05, 1,21
Tasks: 200 total,   1 running, 199 sleeping,   0 stopped,   0 zombie
%Cpu(s):  5,6 us,  1,1 sy,  0,0 ni, 91,2 id,  2,0 wa,  0,0 hi,  0,0 si,  0,0 st
KiB Mem:   4036544 total,  2896804 used,  1139740 free,    22756 buffers
KiB Swap:  4036512 total,   195620 used,  3840892 free.  1539672 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
29696 jack      20   0 1385076 140776  57388 S   0,0  3,5   0:17.05 chromium
 3893 jack      20   0  911900 129744  46412 S   0,0  3,2   2:06.43 chromium
 3863 jack      20   0 3616964 110788  43016 S   0,0  2,7   1:08.25 chromium
 3908 jack      20   0 1284284  71168  12684 S   0,0  1,8   0:24.67 chromium
 3913 jack      20   0 1330884  48564  17188 S   0,0  1,2   0:06.81 chromium
 3934 jack      20   0 1238804  26436  12892 S   0,0  0,7   0:17.35 chromium
 3917 jack      20   0 1235192  24376  12620 S   0,0  0,6   0:08.47 chromium
 2170 root      20   0  184532  22840  16040 S   0,0  0,6   2:58.84 Xorg
 3921 jack      20   0 1230692  19852   9924 S   0,0  0,5   0:06.52 chromium
 2346 jack      20   0  209376  14268   8604 S   0,0  0,4 104:42.91 urxvtd
 1929 bind      20   0  739060  11196   1816 S   0,0  0,3   0:02.12 named
 2273 jack      20   0  190736   9896   5476 S   0,0  0,2   6:59.27 openbox
 3963 jack      20   0 1348092   9304   7432 S   0,0  0,2   0:02.75 chromium
 3925 jack      20   0 1231788   8404   4900 S   0,0  0,2   0:00.36 chromium
31225 jack      20   0   41304   7324   4856 S   0,0  0,2   0:00.06 vi
30484 jack      20   0   24928   6896   3448 S   0,0  0,2   0:00.27 bash
28463 jack      20   0   25000   6752   3392 S   0,0  0,2  42:00.20 bash
29912 jack      20   0   24828   6668   3328 S   0,0  0,2   0:00.24 bash
30333 jack      20   0   24828   6664   3320 S   0,0  0,2   0:25.22 bash
30122 jack      20   0   24820   6520   3220 S   0,0  0,2   0:00.07 bash
 3901 jack      20   0  488808   5972   5468 S   0,0  0,1   0:00.00 chromium
 2324 jack      20   0  144212   5456   4484 S   0,0  0,1   0:09.73 tint2
 3869 jack      20   0  402092   4964   3504 S   0,0  0,1   0:00.04 chromium
```

After:
```
18% [jack:~]top -Sbn1 -o RES | head -30
top - 23:47:01 up  5:36,  6 users,  load average: 1,48, 5,68, 5,65
Tasks: 193 total,   1 running, 192 sleeping,   0 stopped,   0 zombie
%Cpu(s):  5,5 us,  1,4 sy,  0,0 ni, 88,4 id,  4,6 wa,  0,0 hi,  0,0 si,  0,0 st
KiB Mem:   4036544 total,  3312796 used,   723748 free,    10644 buffers
KiB Swap:        0 total,        0 used,        0 free.    82016 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 3863 jack      20   0 3605384 128648  29052 S   0,0  3,2   4:55.87 chromium
 2170 root      20   0  184916  49416  14268 S   0,0  1,2   4:06.03 Xorg
 1929 bind      20   0  739580  32404      0 S   0,0  0,8   0:02.88 named
 2346 jack      20   0  209376  18776   3092 S   0,0  0,5 147:32.83 urxvtd
 2273 jack      20   0  190736   9248   3380 S   0,0  0,2   6:59.74 openbox
 3871 jack      20   0  549556   7008      0 S   0,0  0,2   5:41.88 chromium
 3869 jack      20   0  402092   6968      0 S   0,0  0,2   0:00.04 chromium
 2324 jack      20   0  144212   6872   4944 S   0,0  0,2   0:10.19 tint2
29912 jack      20   0   24828   5808   2468 S   0,0  0,1   0:00.50 bash
30333 jack      20   0   24828   4928   1584 S   0,0  0,1   6:46.61 bash
30484 jack      20   0   24928   3456      8 S   0,0  0,1   0:00.27 bash
31717 jack      20   0   24824   3348      8 S   0,0  0,1   0:00.14 bash
30122 jack      20   0   24820   3300      0 S   0,0  0,1   0:00.07 bash
32539 jack      20   0   23512   2820   2416 R   6,3  0,1   0:00.02 top
31225 jack      20   0   41304   2476      8 S   0,0  0,1   0:00.08 vi
30266 jack      20   0   24488   2028   1252 S   0,0  0,1   0:07.53 htop
 2325 jack      20   0   60636   1864   1100 S   0,0  0,0   0:03.06 xscreensaver
32540 jack      20   0    6204   1648   1556 S   0,0  0,0   0:00.00 head
 2345 jack      20   0   83472   1612   1032 S   0,0  0,0   0:00.02 urxvtd
 2263 root      20   0   84000   1340      8 S   0,0  0,0   0:00.17 xdm
  380 root      20   0   41680   1320      4 S   0,0  0,0   0:00.89 udevd
 2340 jack      20   0  112636   1200    436 S   0,0  0,0   0:02.60 redshift
 1873 root      20   0  258664    900      0 S   0,0  0,0   0:00.40 rsyslogd
```

free -m:
```
10% [jack:/tmp]free -m
             total       used       free     shared    buffers     cached
Mem:          3941       3415        526         25         14        162
-/+ buffers/cache:       3238        703
Swap:            0          0          0
```

**Comment by markfasheh** [comment]
Ok, so I did the run on my test machine, ctrl-c'd it when memory was just about full.

ipcs -p showed no difference before or after - we're not leaking shared memory.

/proc/meminfo before and after tells us some more information:
MemTotal:       16407076 kB
MemFree:        15582272 kB
MemAvailable:   15946916 kB
Buffers:           81716 kB
Cached:           423252 kB
SwapCached:            0 kB
Active:           266104 kB
Inactive:         323260 kB
Active(anon):      85320 kB
Inactive(anon):      552 kB
Active(file):     180784 kB
Inactive(file):   322708 kB


After:
MemTotal:       16407076 kB
MemFree:          438028 kB
MemAvailable:   15885520 kB
Buffers:           82772 kB
Cached:         15398552 kB
SwapCached:            0 kB
Active:          4067668 kB
Inactive:       11500888 kB
Active(anon):        876 kB
Inactive(anon):    88060 kB
Active(file):    4066792 kB
Inactive(file): 11412828 kB

so there's an 11 gig difference in inactive file pages before and after. As one would expect, unmounting the btrfs file system cleans it all up and gives me my memory back:

weyoun2:~ # free -m
             total       used       free     shared    buffers     cached
Mem:         16022        804      15218          1         81        459
-/+ buffers/cache:        264      15758
Swap:         2053          0       2053

I should've gotten vmstat info, I'll do it next time. Also next time I'll do echo 3 > /proc/sys/vm/drop_caches instead of unmount (my guess is that it will work).

So this is probably file data that the dedupe ioctl is pulling in for its comparison. Why it's not being reclaimed (and instead we're hitting oom-killer) is something I need to figure out.

btw, this is all on 3.18

**Comment by dioni21** [comment]
@markfasheh In your example, is the perfect example of why "free memory" is not reliable on Linux.

Note the values:
Before:
Cached: 423252 kB
After:
Cached: 15398552 kB

Your memory is "not free", but it is also available.  It is inactive, but its content can still be used as cache.

"sudo sysctl vm.drop_caches=3" will get all that back, but do not do this except for debugging.

In short: no problem at all...

@clobrother You problem is a little different.

Note that your whole swap area has disappeared on the "after" output. **WHY**?

Anyway, the values do not sum correctly, almost 2G have disappeared.  Please send also the output of "ipcs -p" together with an unlimited version of "top -Sbn1 -o RES" and "cat /proc/meminfo".

**Comment by Enverex** [comment]
Ignoring people misreading actually used vs Cached/Buffered, my issue when I raised this was that it was genuinely holding on to RAM. That said, I've not seen it happen for a while now even after being killed by OOM.

**Comment by clobrother** [comment]
@dioni21: I removed swap to reach OOM condition faster, was late

There are more complete data, drop caches do not change the memory usage, I cannot umount the fs (both lsof & fuser tell me that no processes are in it)
```
8% [jack:~]top -Sbn1 -o RES | head -30
top - 17:00:13 up 6 min,  5 users,  load average: 0,56, 0,66, 0,36
Tasks: 196 total,   1 running, 195 sleeping,   0 stopped,   0 zombie
%Cpu(s):  1,9 us,  1,0 sy,  0,0 ni, 91,5 id,  5,6 wa,  0,0 hi,  0,0 si,  0,0 st
KiB Mem:   4036544 total,  1542204 used,  2494340 free,    48220 buffers
KiB Swap:  4036512 total,        0 used,  4036512 free.   530980 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 2660 jack      20   0 3675896 178452  97452 S   0,0  4,4   0:15.07 chromium
 2691 jack      20   0  898284 178404  87108 S   0,0  4,4   0:20.31 chromium
 2703 jack      20   0 1282384 113780  44256 S   0,0  2,8   0:03.02 chromium
 2734 jack      20   0 1355752 109228  59924 S   0,0  2,7   0:07.14 chromium
 2751 jack      20   0 1330712  91616  50100 S   0,0  2,3   0:01.45 chromium
 2738 jack      20   0 1345008  90856  55280 S   0,0  2,3   0:01.14 chromium
 2170 root      20   0  185376  73624  37596 S   0,0  1,8   0:05.96 Xorg
 2718 jack      20   0 1238584  68376  44516 S   0,0  1,7   0:02.24 chromium
 2706 jack      20   0 1233944  66032  44340 S   0,0  1,6   0:01.52 chromium
 2709 jack      20   0 1230676  60976  43196 S   0,0  1,5   0:00.94 chromium
 2712 jack      20   0 1231168  59392  41836 S   0,0  1,5   0:00.26 chromium
 2666 jack      20   0  402092  36092  29128 S   0,0  0,9   0:00.12 chromium
 1929 bind      20   0  735680  33708   5872 S   0,0  0,8   0:00.35 named
 2696 jack      20   0  488808  27524   7200 S   0,0  0,7   0:00.00 chromium
 2358 jack      20   0  191012  22056  14940 S   0,0  0,5   0:01.19 urxvtd
 2283 jack      20   0  188780  17580  13624 S   0,0  0,4   0:00.67 openbox
 2334 jack      20   0  143880  12228  10600 S   0,0  0,3   0:00.29 tint2
 2668 jack      20   0  549556  12000   5004 S   0,0  0,3   0:13.98 chromium
 2352 jack      20   0  112636   7104   6376 S   0,0  0,2   0:00.14 redshift
 3150 jack      20   0   24828   6748   3408 S   0,0  0,2   0:08.16 bash
 3401 jack      20   0   24824   6716   3380 S   0,0  0,2   0:00.10 bash
 2506 jack      20   0   24820   6648   3340 S   0,0  0,2   0:00.05 bash
 2359 jack      20   0   24824   6628   3336 S   0,0  0,2   0:00.10 bash

8% [jack:~]top -Sbn1 -o RES | head -30
top - 17:26:53 up 33 min,  8 users,  load average: 8,10, 7,95, 6,60
Tasks: 203 total,   1 running, 202 sleeping,   0 stopped,   0 zombie
%Cpu(s):  3,1 us,  5,1 sy,  0,0 ni, 53,7 id, 37,9 wa,  0,0 hi,  0,1 si,  0,0 st
KiB Mem:   4036544 total,  3932668 used,   103876 free,     3768 buffers
KiB Swap:  4036512 total,   270256 used,  3766256 free.   102380 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 8581 jack      20   0  796816  86244  57028 S   0,0  2,1   0:00.70 chromium
 2660 jack      20   0 3668932  57804  26912 S   0,0  1,4   1:04.43 chromium
 8583 jack      20   0  488808  27456   7132 S   0,0  0,7   0:00.00 chromium
 2170 root      20   0  181156  20852  15168 S   0,0  0,5   1:21.94 Xorg
 4654 jack      20   0 1189136  15112   5872 S   6,2  0,4   0:39.15 transmissio+
 7992 jack      20   0  591036  10648   7152 S   0,0  0,3   0:31.06 mplayer
 2358 jack      20   0  207980   5976   2476 S   6,2  0,1   1:16.74 urxvtd
 2283 jack      20   0  191392   5192   2416 S   0,0  0,1   0:02.59 openbox
 2334 jack      20   0  145628   4332   3656 S   0,0  0,1   0:01.05 tint2
 8693 jack      20   0   23648   2748   2332 R   6,2  0,1   0:00.01 top
 3150 jack      20   0   24828   2344   1860 S   0,0  0,1   8:37.08 bash
 3401 jack      20   0   24824   2076   1796 S   0,0  0,1   0:00.15 bash
 8694 jack      20   0    6204   1712   1620 S   0,0  0,0   0:00.00 head
 2335 jack      20   0   60636   1352   1196 S   0,0  0,0   0:00.54 xscreensaver
 1873 root      20   0  258664   1136    564 S   0,0  0,0   0:00.12 rsyslogd
 2352 jack      20   0  112636   1096   1028 S   0,0  0,0   0:00.33 redshift
 7993 jack      20   0  560968    872    844 S   0,0  0,0   0:00.60 mplayer
 2668 jack      20   0  549556    528    476 S   0,0  0,0   0:56.55 chromium
 2068 root      20   0   27496    408    364 S   0,0  0,0   0:02.78 cron
 2503 jack      20   0   44412     52      0 S   0,0  0,0   0:00.32 ssh
 2650 jack      20   0   44412     44      0 S   0,0  0,0   0:00.28 ssh
 3956 jack      20   0   44412     40      0 S   0,0  0,0   0:03.87 ssh
 3812 jack      20   0   24828     16     16 S   0,0  0,0   0:00.07 bash

101% [jack:~]ipcs --all
------ Message Queues --------
key        msqid      owner      perms      used-bytes   messages    

------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status      
0x00000000 1540096    jack       600        393216     2          dest         
0x00000000 11304961   jack       600        3136360    2          dest         
0x00000000 11337730   jack       600        420024     2          dest         
0x0056a4d5 24412163   jack       660        488        1                       
0x0056a4d6 24444932   jack       660        65536      1                       
0x00000000 27295749   jack       600        393216     2          dest         

------ Semaphore Arrays --------
key        semid      owner      perms      nsems     
0x0056a4d5 131072     jack       660        1         

42% [jack:~]ipcs -p
------ Message Queues PIDs --------
msqid      owner      lspid      lrpid     

------ Shared Memory Creator/Last-op PIDs --------
shmid      owner      cpid       lpid      
1540096    jack       2660       8747      
11304961   jack       4654       2170      
11337730   jack       4654       2170      
24412163   jack       7992       7992      
24444932   jack       7992       7992      
27295749   jack       2660       8747      


35% [jack:~]free -m
             total       used       free     shared    buffers     cached
Mem:          3941       3832        109          9          1         62
-/+ buffers/cache:       3769        172
Swap:         3941        363       3578

29% [jack:~]echo 3 | sudo tee /proc/sys/vm/drop_caches
25% [jack:~]top -Sbn1 -o RES | head -30
top - 17:30:47 up 37 min,  9 users,  load average: 2,02, 5,14, 5,79
Tasks: 202 total,   1 running, 201 sleeping,   0 stopped,   0 zombie
%Cpu(s):  2,9 us,  4,7 sy,  0,0 ni, 57,3 id, 35,0 wa,  0,0 hi,  0,1 si,  0,0 st
KiB Mem:   4036544 total,  3917428 used,   119116 free,     5856 buffers
KiB Swap:  4036512 total,   377332 used,  3659180 free.    60972 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 8581 jack      20   0  870756  68812  26088 S   0,0  1,7   0:04.16 chromium
 8746 jack      20   0 1341124  31244  14076 S   0,0  0,8   0:04.08 chromium
 2170 root      20   0  211108  28624  11832 S   0,0  0,7   1:28.50 Xorg
 2660 jack      20   0 3746856  19948   7532 S   0,0  0,5   1:08.87 chromium
 4654 jack      20   0 1189136  13752   3624 S   6,3  0,3   0:43.70 transmissio+
 2358 jack      20   0  211644   9124   2308 S   0,0  0,2   1:21.54 urxvtd
 9714 jack      20   0   41256   7240   4812 S   0,0  0,2   0:00.08 vi
 2283 jack      20   0  192568   5552   1872 S   0,0  0,1   0:02.90 openbox
 8583 jack      20   0  488808   4680   4400 S   0,0  0,1   0:00.00 chromium
 3401 jack      20   0   25012   4336   2676 S   0,0  0,1   0:00.71 bash
 8836 jack      20   0   24932   3680   1280 S   0,0  0,1   0:01.02 bash
 2334 jack      20   0  145628   2972   2356 S   0,0  0,1   0:01.30 tint2
 9767 jack      20   0   23644   2804   2392 R   6,3  0,1   0:00.01 top
 3150 jack      20   0   24844   2036   1324 S   0,0  0,1   8:37.35 bash
 9768 jack      20   0    6204   1692   1600 S   0,0  0,0   0:00.00 head
 7992 jack      20   0  591036   1244   1168 S   0,0  0,0   0:31.65 mplayer
 2335 jack      20   0   60636    920    732 S   0,0  0,0   0:00.67 xscreensaver
 2068 root      20   0   27496    552    468 S   0,0  0,0   0:03.24 cron
 2357 jack      20   0   83472    384    316 S   0,0  0,0   0:00.01 urxvtd
 9662 jack      20   0   29004    332      0 S   0,0  0,0   0:00.00 xclip
 2503 jack      20   0   44412    152     52 S   0,0  0,0   0:00.36 ssh
 2650 jack      20   0   44412    144    136 S   0,0  0,0   0:00.28 ssh
 1873 root      20   0  258664    132      0 S   0,0  0,0   0:00.15 rsyslogd

21% [jack:~]free -m
             total       used       free     shared    buffers     cached
Mem:          3941       3833        108          7          1         71
-/+ buffers/cache:       3760        181
Swap:         3941        366       3575
```

**Comment by markfasheh** [comment]
dioni21: I understand that this isn't a memory leak. What I'm saying is that we shouldn't be ooming, especially considering that all the memory in question is freeable. There's either something wrong with what duperemove is doing or what the btrfs-extent-same ioctl code is doing. I haven't figured out which one it is yet.

**Comment by markfasheh** [comment]
https://github.com/Enverex: yeah I don't think we've been talking about your original bug for a while now. If all the traffic is bothersome to you we can move this to another issue.

**Comment by Enverex** [comment]
No worries, I was just making sure people didn't confuse that issue with any others.

**Comment by dioni21** [comment]
@clobrother your problem appears to be memory allocated in kernel!

You can debug some of that by looking at the contents of **/proc/slabinfo** or running **slabtop**
I just found this page: https://www.kernel.org/doc/Documentation/kmemleak.txt

Other than that, I don't think I can be of more help.  Debugging kernel memory is very specific, and rarely needed.

The fact that you cannot unmount the fs may infer that something is already terribly wrong.  Did you reboot since the first occurrence?

**Comment by dioni21** [comment]
@markfasheh OOM can happen everytime you have less RAM than needed.

I have personally OOMed my system by running dupremove in a large file system using a very small block size.

What is **not** normal in this case is the memory to be kept in use **after** the program has exited.

**Comment by clobrother** [comment]
@dioni21 interesting stuff, thank
I reboot after each test (because I have no more memory!)
The "something" which hold the FS reminds me two others things : drbd & NFS. Both are in kernel space, nothing in userspace

slabtop does not show anything special, there is one after a test:
```
 Active / Total Objects (% used)    : 203040 / 251639 (80,7%)
 Active / Total Slabs (% used)      : 17551 / 17573 (99,9%)
 Active / Total Caches (% used)     : 140 / 244 (57,4%)
 Active / Total Size (% used)       : 64654,78K / 77072,27K (83,9%)
 Minimum / Average / Maximum Object : 0,02K / 0,31K / 4096,00K

  OBJS ACTIVE  USE OBJ SIZE  SLABS OBJ/SLAB CACHE SIZE NAME                   
 22330  11894  53%    0,56K   3190        7     12760K radix_tree_node        
 14077  13753  97%    0,56K   2011        7      8044K inode_cache            
 41643  41643 100%    0,19K   1983       21      7932K dentry                 
   309    309 100%   16,00K    309        1      4944K kmalloc-16384          
  1053   1053 100%    4,00K   1053        1      4212K kmalloc-4096           
  3252   2280  70%    0,96K    813        4      3252K btrfs_inode            
  9581   2627  27%    0,29K    737       13      2948K btrfs_extent_buffer    
 20625  20617  99%    0,12K    625       33      2500K kernfs_node_cache      
  2412   2412 100%    0,98K    603        4      2412K ext4_inode_cache       
  2276   2276 100%    1,00K    569        4      2276K kmalloc-1024           
  3054   2669  87%    0,62K    509        6      2036K proc_inode_cache       
 12908   5338  41%    0,14K    461       28      1844K btrfs_extent_map       
    23     23 100%   64,00K     23        1      1472K kmalloc-65536          
  6972   6890  98%    0,19K    332       21      1328K kmalloc-192            
 10075   6199  61%    0,12K    325       31      1300K kmalloc-128            
   644    644 100%    2,00K    322        2      1288K kmalloc-2048  
```

**Comment by markfasheh** [comment]
Closing, this was fixed in the kernel.
